warmrestart/warm_restart.cpp

```diff
@@ -0,0 +1,105 @@
+#include <string>
+#include <vector>
+#include "logger.h"
+#include "schema.h"
+#include "warm_restart.h"
+
+namespace swss {
+
+const WarmStart::WarmStartStateNameMap WarmStart::warmStartStateNameMap =
+{
+    {INIT,          "init"},
+    {RESTORED,      "restored"},
+    {RECONCILED,    "reconciled"}
+};
+
+WarmStart &WarmStart::getInstance()
+{
+    static WarmStart m_warmStart;
+    return m_warmStart;
+}
+
+// Check warm start flag at the very begining of application, do it once for each process.
+bool WarmStart::checkWarmStart(const std::string &app_name, const std::string &docker_name)
+{
+    auto& warmStart = getInstance();
+
+    if (warmStart.m_stateDb)
+    {
+        return true;
+    }
+
+    warmStart.m_stateDb          = std::make_shared<swss::DBConnector>(STATE_DB, swss::DBConnector::DEFAULT_UNIXSOCKET, 0);
+    warmStart.m_stateRedisClient = std::make_shared<swss::RedisClient>(warmStart.m_stateDb.get());
+
+    warmStart.m_cfgDb          = std::make_shared<swss::DBConnector>(CONFIG_DB, swss::DBConnector::DEFAULT_UNIXSOCKET, 0);
+    warmStart.m_cfgRedisClient = std::make_shared<swss::RedisClient>(warmStart.m_cfgDb.get());
+
+    bool system_warm_start = false;
+    warmStart.enabled = false;
+
+    // Check system level warm restart config first
+    auto pvalue = warmStart.m_cfgRedisClient->hget(CFG_WARM_RESTART_TABLE_NAME"|system", "enable");
```

Comment:
| should not be hard coded.

Reply:
use table api?

Reply:
Probably table should be expanded with member functions like getEntry(const std::string &key, const std::string &field) and setEntry(const std::string &key, const std::string &field, const std::string &value)

I switched to using RedisClient to avoid the clumsy key level only operations.

Will give it a try.